### ISSUE TITLE
Technical debt reduction items

### ISSUE THREAD

**Issue by marcocitus** [citusdata/citus#1848]
Projects that reduce technical debt and make Citus code easier to understand:

- [x] Unified executor (#1273): Get rid of router / multi-shard / real-time / task-tracker / other executors and provide one code path instead.

- [x] Query pushdown (#1706): Get rid of fixed join order planner.

- [x] Fix OR shard pruning bug (#1537): This code is fast, but hard to understand. We may want to make it more understandable as part of fixing the OR problem.

- [x] Implement prepared statement support: We currently have 4 different ways of handling prepared statements and they mostly fail to provide prepared statement benefits. We should be able to reason about query parameters (e.g. does this query prune down to 1 shard once the parameter is known?).

- [x] Merge shard rebalancer extension into Citus enterprise.

- [x] Improve code organisation, especially within the planner. The boundaries between the logical planner, optimiser, physical planner, subquery pushdown planner, router planner, and INSERT...SELECT planner have diffused over time and functions are scattered across files. It makes the code harder to understand.

**Comment by ozgune** [comment]
I'll add a few items to this thread. These items made us waste notable amounts of time as we were trying Citus on different platforms:

* `FinishConnectionEstablishment()` says that it synchronously closes a connection. However, the function loops until a connection is established and polls the socket. In general, most functions related to `connection_management.c` were hard to follow. When opening one connection, one needs to go down 9-10 levels down the call stack. Why can't we just call the related libpq functions? -- don't know, just asking

* We explicitly allocate and use `LWLockTranche`. Example code from pg_stat_statements is simpler and doesn't make those extra allocations. We can significantly simplify our code by following what pg_stat_statements does.

**Comment by marcocitus** [comment]
> Why can't we just call the related libpq functions?

Those don't support cancellation, but I agree that functions like `FinishConnectionEstablishment` are convoluted.

**Comment by marcocitus** [comment]
One more item that came up:

- Remove distinction between original query and rewritten query in logical planner. Subquery pushdown uses original query, logical planner uses rewritten query. The code would be much simpler if we just used one (probably original).

**Comment by marcocitus** [comment]
Most of these items are addressed.
